### ISSUE TITLE
Design flaws around status register and error flags

### ISSUE THREAD

**Issue by YruamaLairba** [samcrow/stm32_i2s#4]
Hi, in fact several issues

The first issue is related to side effects occurring when reading SR and DR registers:
 - underrun and frame error flag are deleted when reading SR register
 - overrun flag is deleted when reading DR and then SR registers

In other words, it change the internal state of the device. From API design point of view that means all methods triggering those side effects should take mutably a reference to the hal and should documents those side effects.

The second issue is about ergonomy: i find the type returned by `take_error()` unergonomic and probably inefficient. For example let say i want to check for frame error, i need to do 3 test:
```rust
if let Err(e) = i2s.take_error(){
    match e {
        Frame | FrameAndOverrun(_) => {/*handle the error*/},
        _ => (),
    }
}
```
i think we could replace the `Result<(), ReceiveError>` by an opaque type allowing this
```rust
if i2s.take_error().frame_error(){
    /*handle the error*/
}
```
And Frame error is missing from TransmitError

**Comment by samcrow** [comment]
1. You're right, and I would welcome a pull request to fix this.
2. I don't want to completely give up on using a Result, but I agree that it would be good to change ReceiveError (maybe even making it opaque with no visible enum variants) to make it easier to use.

**Comment by YruamaLairba** [comment]
from my point of view, having `take_error() -> Result<(), Error>` is even debatable from semantic point of view. From my point of view, returning a `Result<_,Error>` means Errors can occur when doing this operation. Here instead, those Error occurred before the `take_error()` call, and the method actually just retrieve those Error. With that in mind and from a purely semantic point of view the API should be `take_error() -> Option<Errors>`.

Anyway:
 - Before doing anything, this worth to see in other HAL to see how similar case are handled. Personnally, my approach would be more raw, I would probably using a `status() -> Status` method
 - i'm busy working on duplex i2s support. I may fix here only things preventing me to make it work.

**Comment by samcrow** [comment]
I now agree that having the function return an option of an error type is another good approach.

**Comment by YruamaLairba** [comment]
Closing, this is solved by the redesign done in 0.3
